// zod/index.js
class ZodType {
  constructor(parser = (value) => value) {
    this._parser = parser;
  }

  parse(value) {
    return this._parser(value);
  }

  safeParse(value) {
    try {
      return { success: true, data: this.parse(value) };
    } catch (error) {
      return { success: false, error };
    }
  }

  optional() {
    const parent = this;
    return new ZodType((value) => {
      if (value === undefined) {
        return undefined;
      }
      return parent.parse(value);
    });
  }

  nullish() {
    const parent = this;
    return new ZodType((value) => {
      if (value == null) {
        return value;
      }
      return parent.parse(value);
    });
  }

  nullable() {
    return this.nullish();
  }

  min() {
    return this;
  }

  url() {
    return this;
  }

  default(defaultValue) {
    const parent = this;
    return new ZodType((value) => {
      if (value === undefined) {
        return typeof defaultValue === 'function' ? defaultValue() : defaultValue;
      }
      return parent.parse(value);
    });
  }

  transform(fn) {
    const parent = this;
    return new ZodType((value) => fn(parent.parse(value)));
  }

  passthrough() {
    return this;
  }

  partial() {
    return this;
  }

  strict() {
    return this;
  }
}

const z = {
  string() {
    return new ZodType((value) => (typeof value === 'string' ? value : String(value ?? '')));
  },
  number() {
    return new ZodType((value) => {
      if (typeof value === 'number') {
        return value;
      }
      const parsed = Number(value);
      if (Number.isNaN(parsed)) {
        throw new Error('Expected number');
      }
      return parsed;
    });
  },
  boolean() {
    return new ZodType((value) => Boolean(value));
  },
  null() {
    return new ZodType(() => null);
  },
  unknown() {
    return new ZodType((value) => value);
  },
  array(schema) {
    return new ZodType((value) => {
      if (!Array.isArray(value)) {
        throw new Error('Expected array');
      }
      return schema ? value.map((item) => schema.parse(item)) : value.slice();
    });
  },
  enum(values) {
    return new ZodType((value) => {
      const normalized = typeof value === 'string' ? value : String(value ?? '');
      if (!values.includes(normalized)) {
        throw new Error(`Invalid enum value: ${normalized}`);
      }
      return normalized;
    });
  },
  union(schemas) {
    return new ZodType((value) => {
      for (const schema of schemas) {
        try {
          return schema.parse(value);
        } catch (error) {
          // continue
        }
      }
      throw new Error('Value did not match any union member');
    });
  },
  record(keySchema, valueSchema = z.unknown()) {
    return new ZodType((value) => {
      const result = {};
      if (!value || typeof value !== 'object') {
        return result;
      }
      for (const [key, entry] of Object.entries(value)) {
        const parsedKey = keySchema ? keySchema.parse(key) : key;
        result[parsedKey] = valueSchema ? valueSchema.parse(entry) : entry;
      }
      return result;
    });
  },
  object(shape = {}) {
    return new ZodType((value) => {
      const result = {};
      if (!value || typeof value !== 'object') {
        return result;
      }
      for (const [key, schema] of Object.entries(shape)) {
        if (schema instanceof ZodType) {
          const parsed = schema.parse(value[key]);
          if (parsed !== undefined) {
            result[key] = parsed;
          }
        } else if (schema && typeof schema.parse === 'function') {
          const parsed = schema.parse(value[key]);
          if (parsed !== undefined) {
            result[key] = parsed;
          }
        } else {
          result[key] = value[key];
        }
      }
      return result;
    });
  },
};

z.nullish = () => z.null().optional();
z.literal = (value) => new ZodType((input) => (input === value ? input : value));

module.exports = { z, default: z };
